frame/staking/src/lib.rs

```diff
@@ -846,8 +1005,77 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		fn on_initialize() {
+		/// Does the following:
+		///
+		/// 1. potential storage migration
+		/// 2. sets `ElectionStatus` to `Triggered(now)` where `now` is the block number at which
+		///    the election window has opened. The offchain worker, if applicable, will execute at
+		///    the end of the current block. `submit_election_solution` will accept solutions from
+		///    this block until the end of the era.
+		fn on_initialize(now: T::BlockNumber) {
 			Self::ensure_storage_upgraded();
+			if
+				// if we don't have any ongoing offchain compute.
+				Self::era_election_status() == ElectionStatus::Close &&
+				// and an era is about to be changed.
+				Self::is_current_session_final()
+			{
+				let next_session_change =
+					T::NextSessionChange::estimate_next_session_change(now);
+				if let Some(remaining) = next_session_change.checked_sub(&now) {
+					if remaining <= T::ElectionLookahead::get() && !remaining.is_zero() {
+						// create snapshot.
+						if Self::create_stakers_snapshot() {
+							// Set the flag to make sure we don't waste any compute here in the same era
+							// after we have triggered the offline compute.
+							<EraElectionStatus<T>>::put(
+								ElectionStatus::<T::BlockNumber>::Open(now)
+							);
+							debug::native::info!(
+								target: "staking",
+								"Election window is Open({:?}). Snapshot created",
+								now,
+							);
+						} else {
+							debug::native::warn!(
+								target: "staking",
+								"Failed to create snapshot at {:?}. Election window will remain closed.",
+								now,
+							);
+						}
+
+					}
+				}
+			}
+		}
+
+		/// Check if the current block number is the one at which the election window has been set
+		/// to open. If so, it runs the offchain worker code.
+		fn offchain_worker(now: T::BlockNumber) {
+			debug::RuntimeLogger::init();
+			use offchain_election::{set_check_offchain_execution_status, compute_offchain_election};
+
+			let window_open =
+				Self::era_election_status() == ElectionStatus::<T::BlockNumber>::Open(now);
+
+			if window_open {
+				let offchain_status = set_check_offchain_execution_status::<T>(now);
+				if let Err(why) = offchain_status {
+					debug::native::warn!(
+						target: "staking",
+						"skipping offchain call in open election window due to [{}]",
```

Comment:
This warning will be pretty spammy, since it will get printed if the offchain calculation was recently executed. I think it would be best to tone down that special case a bit.

Reply:
It would be once per era at most, right? since we are behind an `if` to ensure we run only when the window is open `now`.

Reply:
This raises another concern if it should be 

```
let window_open =
				Self::era_election_status() == ElectionStatus::<T::BlockNumber>::Open(now);
```

or 

```
let window_open =
				Self::era_election_status() == ElectionStatus::<T::BlockNumber>::Open(...);
```

I don't see a reason why to go with the latter. This is guaranteed to be `true` at least once per era, which is exactly what we want.

Reply:
Well the on-chain state will only change if the transaction ends up updating it, right? So if it takes more than one block to produce the results + some additional blocks to actually get the transaction included we may end up running a bunch of offchain workers in the meantime that will fail with the warning, no?

Regarding the second question: The later indeed makes sense too, the former is just more specific. I don't mind either.

Reply:
> Well the on-chain state will only change if the transaction ends up updating it, right?

The state that indicates the window is open, no. That one is set independently, on chain. see `EraElectionStatus` being set in `on_initialise()`